src/cephadm/cephadmlib/listing.py

```diff
@@ -0,0 +1,333 @@
+                if '-' not in entry:
+                    continue  # invalid entry
+                (cluster, daemon_id) = entry.split('-', 1)
+                fsid = get_legacy_daemon_fsid(
+                    ctx,
+                    cluster,
+                    daemon_type,
+                    daemon_id,
+                    legacy_dir=legacy_dir,
+                )
+                legacy_unit_name = 'ceph-%s@%s' % (daemon_type, daemon_id)
+                yield LegacyDaemonEntry(
+                    fsid=fsid or '',
+                    daemon_type=daemon_type,
+                    name=legacy_unit_name,
+                    status={
+                        'style': LEGACY,
+                        'name': '%s.%s' % (daemon_type, daemon_id),
+                        'fsid': fsid if fsid is not None else 'unknown',
+                        'systemd_unit': legacy_unit_name,
+                    },
+                    data_dir=data_dir,
+                )
+        elif is_fsid(dirname):
+            assert isinstance(dirname, str)
+            fsid = dirname
+            cluster_dir = os.path.join(data_dir, fsid)
+            for entry in os.listdir(cluster_dir):
+                if not (
+                    '.' in entry
+                    and os.path.isdir(os.path.join(cluster_dir, entry))
+                ):
+                    continue  # invalid entry
+                identity = DaemonIdentity.from_name(fsid, entry)
+                yield DaemonEntry(
+                    identity=identity,
+                    status={
+                        'style': VERSION1,
+                        'name': identity.daemon_name,
+                        'fsid': fsid,
+                        'systemd_unit': identity.unit_name,
+                    },
+                    data_dir=data_dir,
+                )
+
+
+def daemons_matching(
+    ctx: CephadmContext,
+    legacy_dir: Optional[str] = None,
+    daemon_name: Optional[str] = None,
+    daemon_type: Optional[str] = None,
+    fsid: Optional[str] = None,
+) -> Iterator[Union[LegacyDaemonEntry, DaemonEntry]]:
+    """Iterate over the daemons configured on the current node, matching daemon
+    name or daemon type if supplied.
+    """
+    for entry in daemons(ctx, legacy_dir):
+        if isinstance(entry, LegacyDaemonEntry):
+            if fsid is not None and fsid != entry.fsid:
+                continue
+            if daemon_type is not None and daemon_type != entry.daemon_type:
+                continue
+        elif isinstance(entry, DaemonEntry):
+            if fsid is not None and fsid != entry.identity.fsid:
+                continue
+            if (
+                daemon_name is not None
+                and daemon_name != entry.identity.daemon_name
+            ):
+                continue
+            if (
+                daemon_type is not None
+                and daemon_type != entry.identity.daemon_type
+            ):
+                continue
```

Comment:
Cosmetic: since the checks are very similar, consider having a helper func such as:

```
def field_mismatch(filter_value, actual_value) -> bool:
    return filter_value is not None and filter_value != actual_value
```

So the code becomes something like:
```
if field_mismatch(fsid, entry.fsid):
    continue
```
This way we can put debugging inside the function (if we need to) easily ...

Reply:
I'll consider it for the next time I touch this code, thanks.